### PR TITLE
feat(swarm): implement `Copy` and `Clone` for `FromSwarm`

### PR DESCRIPTION
**Pull request by thomaseizinger**
## Description

We can make `FromSwarm` implement `Copy` and `Close` which makes it much easier to

a) generate code in `libp2p-swarm-derive`
b) manually wrap a `NetworkBehaviour`

Previously, we couldn't do this because `ConnectionClosed` would have a `handler` field that cannot be cloned / copied.

Related: #4076.
Related: #4581.

## Change checklist

- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] A changelog entry has been made in the appropriate crates

### CODE REVIEW

**Review by nazar-pc — APPROVED** [review]
This is much nicer than before

**Comment by thomaseizinger** [comment]
Apply `internal-change` for `autonat` and `request-response`.

**Comment by mxinden** [comment]
:+1: for `Clone`.

`Copy`: What if we want to add a non-copy enum variant in the future? That would be a breaking change? Is that worth the ergonomics through `Copy`?

**Comment by thomaseizinger** [comment]
> 👍 for `Clone`.
> 
> `Copy`: What if we want to add a non-copy enum variant in the future? That would be a breaking change? Is that worth the ergonomics through `Copy`?

At the moment, all non-Copy types in the enum variants are references which are Copy. I think long-term, we might want to pass a reference of the entire `FromSwarm` enum down instead such that we can return the actual `FromSwarm` instance as part of `SwarmEvent`.

**Comment by thomaseizinger** [comment]
> > 👍 for `Clone`.
> > `Copy`: What if we want to add a non-copy enum variant in the future? That would be a breaking change? Is that worth the ergonomics through `Copy`?
> 
> At the moment, all non-Copy types in the enum variants are references which are Copy. I think long-term, we might want to pass a reference of the entire `FromSwarm` enum down instead such that we can return the actual `FromSwarm` instance as part of `SwarmEvent`.

We can remove `Copy` again at that point if we want to because users will just receive a reference anyway. Until then, I think `Copy` is the way to go.

**Review by mxinden — COMMENTED** [review]
Thank you for expanding on the reasoning here.

Indeed my historical intuition of `Copy` is wrong. In other words agreed that `Copy` makes sense here.

Agreed with the long term goal to pass an `&FromSwarm` to the `NetworkBehaviour`s and then the same `FromSwarm` to the user in the future.

I believe with this change `libp2p-swarm-derive` can be significantly simplified. I don't think we need the statements below any longer. Mind simplifying this as part of this pull request?

https://github.com/libp2p/rust-libp2p/blob/bb2b79824d1fd9e4f9cf8395ace810d032363397/swarm-derive/src/lib.rs#L224-L536

**Comment by thomaseizinger** [comment]
> I believe with this change `libp2p-swarm-derive` can be significantly simplified. I don't think we need the statements below any longer. Mind simplifying this as part of this pull request?

I already simplified parts of `libp2p-swarm-derive`. Is that what you mean or do you see further potential for improvement?

**Comment by thomaseizinger** [comment]
Friendly ping @mxinden.

File: swarm-derive/CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,8 @@
+## 0.34.1 - unreleased
```
Comment (thomaseizinger) [inline swarm-derive/CHANGELOG.md]:
```suggestion
## 0.34.1
```

File: swarm/CHANGELOG.md [hunk 2]
```diff
@@ -1,3 +1,9 @@
+## 0.44.1 - unreleased
```
Comment (thomaseizinger) [inline swarm/CHANGELOG.md]:
```suggestion
## 0.44.1
```

**Comment by mxinden** [comment]
> > I believe with this change `libp2p-swarm-derive` can be significantly simplified. I don't think we need the statements below any longer. Mind simplifying this as part of this pull request?
> 
> I already simplified parts of `libp2p-swarm-derive`. Is that what you mean or do you see further potential for improvement?

Seems like I looked at the wrong revision. Sorry. The changes you made were the ones I had in mind.

**Review by mxinden — APPROVED** [review]

**Pull request merged**